test/e2e/storage/drivers/in_tree.go

```diff
@@ -0,0 +1,1388 @@
+
+import (
+	"fmt"
+	"math/rand"
+	"os/exec"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/api/core/v1"
+	rbacv1beta1 "k8s.io/api/rbac/v1beta1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeletapis "k8s.io/kubernetes/pkg/kubelet/apis"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	vspheretest "k8s.io/kubernetes/test/e2e/storage/vsphere"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+// NFS
+type nfsDriver struct {
+	serverIP               string
+	serverPod              *v1.Pod
+	externalProvisionerPod *v1.Pod
+	externalPluginName     string
+
+	driverInfo DriverInfo
+}
+
+// InitNFSDriver returns nfsDriver that implements TestDriver interface
+func InitNFSDriver() TestDriver {
+	return &nfsDriver{
+		driverInfo: DriverInfo{
+			Name:        "nfs",
+			MaxFileSize: testpatterns.FileSizeLarge,
+			SupportedFsType: map[string]bool{
+				"": true, // Default fsType
+			},
+			IsPersistent:       true,
+			IsFsGroupSupported: false,
+			IsBlockSupported:   false,
+		},
+	}
+}
+
+func (n *nfsDriver) GetDriverInfo() *DriverInfo {
+	return &n.driverInfo
+}
+
+func (n *nfsDriver) SkipUnsupportedTest(tp testpatterns.TestPattern) {
+}
+
+func (n *nfsDriver) GetVolumeSource(readOnly bool, fsType string) *v1.VolumeSource {
+	return &v1.VolumeSource{
+		NFS: &v1.NFSVolumeSource{
+			Server:   n.serverIP,
+			Path:     "/",
+			ReadOnly: readOnly,
+		},
+	}
+}
+
+func (n *nfsDriver) GetPersistentVolumeSource(readOnly bool, fsType string) *v1.PersistentVolumeSource {
+	return &v1.PersistentVolumeSource{
+		NFS: &v1.NFSVolumeSource{
+			Server:   n.serverIP,
+			Path:     "/",
+			ReadOnly: readOnly,
+		},
+	}
+}
+
+func (n *nfsDriver) GetDynamicProvisionStorageClass(fsType string) *storagev1.StorageClass {
+	provisioner := n.externalPluginName
+	parameters := map[string]string{"mountOptions": "vers=4.1"}
+	ns := n.driverInfo.Framework.Namespace.Name
+	suffix := fmt.Sprintf("%s-sc", n.driverInfo.Name)
+
+	return getStorageClass(provisioner, parameters, nil, ns, suffix)
+}
+
+func (n *nfsDriver) CreateDriver(vt testpatterns.TestVolType) {
+	f := n.driverInfo.Framework
+	cs := f.ClientSet
+	ns := f.Namespace
+	n.externalPluginName = "example.com/nfs"
+
+	// NewNFSServer creates a pod for InlineVolume and PreprovisionedPV,
+	// and startExternalProvisioner creates a pods for DynamicPV.
+	// Therefore, we need a different CreateDriver logic for volType.
+	switch vt {
+	case testpatterns.InlineVolume:
+		fallthrough
+	case testpatterns.PreprovisionedPV:
+		n.driverInfo.Config, n.serverPod, n.serverIP = framework.NewNFSServer(cs, ns.Name, []string{})
+	case testpatterns.DynamicPV:
+		// TODO(mkimuram): cluster-admin gives too much right but system:persistent-volume-provisioner
```

Comment:
What is missing from persistent-volume-provisioner?  The existing external provisioner test does not need cluster-admin.

Reply:
> What is missing from persistent-volume-provisioner? 

Test fails with below error:

```
      Failed to create injector pod: an empty namespace may not be set during creation
      Expected error:
          <*errors.errorString | 0xc421aa7600>: {
              s: "an empty namespace may not be set during creation",
          }
          an empty namespace may not be set during creation
      not to have occurred
```

And system:persistent-volume-provisioner in my test environment is as follows:
```
# kubectl describe clusterrole system:persistent-volume-provisioner
Name:         system:persistent-volume-provisioner
Labels:       kubernetes.io/bootstrapping=rbac-defaults
Annotations:  rbac.authorization.kubernetes.io/autoupdate=true
PolicyRule:
  Resources                      Non-Resource URLs  Resource Names  Verbs
  ---------                      -----------------  --------------  -----
  persistentvolumes              []                 []              [create delete get list watch]
  persistentvolumeclaims         []                 []              [get list update watch]
  storageclasses.storage.k8s.io  []                 []              [get list watch]
  events                         []                 []              [watch create patch update]
```

While nfs provisioner seems to require more:

https://github.com/kubernetes-incubator/external-storage/blob/master/nfs-client/deploy/auth/clusterrole.yaml

> The existing external provisioner test does not need cluster-admin.

I guess that this issue doesn't appear until the test try to create a pod which attaches the provisioned volume.

I intend to fix it after this PR is merged. Because it is beyond scope of refactoring.
(My intention for adding NFS here was to test dynamic provision code in my test environment, for it was only dynamic provisioner that can be tested without cloud provider.)